### PR TITLE
Support Jenkins-style H (hash) schedules

### LINKED ISSUE #1: Support Jenkins-style H (hash) schedules
We should support Jenkins-style cron expressions where `H` may be given instead of a minute, hour, DoM, month, or DoW. `H` ("hash") means "pick a random value". But it's randomized according to a fixed seed provided by the user, so that a particular kind of job can consistently use the same value. (For instance the seed might be a job name or a database ID.)

Here is the full Jenkins documentation for its own cron syntax:

<details>
  <summary>Expand</summary>

```
This field follows the syntax of cron (with minor differences). Specifically,
each line consists of 5 fields separated by TAB or whitespace:

MINUTE HOUR DOM MONTH DOW
MINUTE	Minutes within the hour (0–59)
HOUR	The hour of the day (0–23)
DOM	The day of the month (1–31)
MONTH	The month (1–12)
DOW	The day of the week (0–7) where 0 and 7 are Sunday.

To specify multiple values for one field, the following operators are available. In the order of precedence,

* specifies all valid values
M-N specifies a range of values
M-N/X or */X steps by intervals of X through the specified range or whole valid range
A,B,...,Z enumerates multiple values

To allow periodically scheduled tasks to produce even load on the system, the
symbol H (for “hash”) should be used wherever possible. For example, using 0 0 * * *
for a dozen daily jobs will cause a large spike at midnight. In contrast,
using H H * * * would still execute each job once a day, but not all at the same
time, better using limited resources.

The H symbol can be used with a range. For example, H H(0-7) * * * means some
time between 12:00 AM (midnight) to 7:59 AM. You can also use step intervals
with H, with or without ranges.

The H symbol can be thought of as a random value over a range, but it actually
is a hash of the job name, not a random function, so that the value remains
stable for any given project.

Beware that for the day of month field, short cycles such as */3 or H/3 will not
work consistently near the end of most months, due to variable month lengths.
For example, */3 will run on the 1st, 4th, …31st days of a long month, then
again the next day of the next month. Hashes are always chosen in the 1-28
range, so H/3 will produce a gap between runs of between 3 and 6 days at the end
of a month. (Longer cycles will also have inconsistent lengths but the effect
may be relatively less noticeable.)

Empty lines and lines that start with # will be ignored as comments.

In addition, @yearly, @annually, @monthly, @weekly, @daily, @midnight, and
@hourly are supported as convenient aliases. These use the hash system for
automatic balancing. For example, @hourly is the same as H * * * * and could
mean at any time during the hour. @midnight actually means some time between
12:00 AM and 2:59 AM.

Examples:

# every fifteen minutes (perhaps at :07, :22, :37, :52)
H/15 * * * *
# every ten minutes in the first half of every hour (three times, perhaps at :04, :14, :24)
H(0-29)/10 * * * *
# once every two hours at 45 minutes past the hour starting at 9:45 AM and finishing at 3:45 PM every weekday.
45 9-16/2 * * 1-5
# once in every two hours slot between 9 AM and 5 PM every weekday (perhaps at 10:38 AM, 12:38 PM, 2:38 PM, 4:38 PM)
H H(9-16)/2 * * 1-5
# once a day on the 1st and 15th of every month except December
H H 1,15 1-11 *
```

</details>

To be concrete, here is the proposed new API.

```
// ParseWithHash is like Parse but additionally supports the symbol H in place
// of the minute, hour, day of month, month, or day of week field. The H symbol
// requests a random value (within the valid range) for each instance of H in
// the cron expression fixed using the given seed.
//
// For example, the schedule
//
//	H H * * *
//
// is a schedule that fires once per day at a random hour and minute that is
// chosen when the schedule is parsed. Given the same input expression and seed,
// the same schedule is generated.
//
// The range for randomly generated day of month values is [1, 28].
//
// Additionally, ParseWithHash interprets the named schedules differently from
// Parse:
//
//   - "@monthly" means "H H H * *"
//   - "@weekly" means "H H * * H"
//   - "@daily" means "H H * * *"
//   - "@hourly" means "H * * * *"
//
// The idea of the H symbol is borrowed from Jenkins, though the details are a
// bit different.
func ParseWithHash(expr string, seed uint64) (*Schedule, error)
```

Assorted notes:

* Things like `H/5 * * * *` should work
* We'll skip implementing `H` with ranges (e.g., `H(0-7)`) that Jenkins supports for now, for simplicity. We can always choose to add it later.
* `Parse` should refuse to parse schedules with `H` (but should give a helpful error pointing out that only `ParseWithHash` handles such schedules).

### PR DESCRIPTION
**Pull request by stefan-toubia**
closes #1

Adds a new `ParseWithHash` function to support "hashed" schedules like `H * * * *`. The `H` symbol is used to randomize scheduled start times to more evenly distribute resource usage, while still maintaining a consistent schedule period (e.g. once per hour).

This implementation adds a `hashedFields` which takes a seed value to generate a deterministic set of random field values that will be used for substituting any `H` values in the schedule expression. This does guarantee some stability in the schedule randomness even when the expression is modified.

For example, a schedule like `0 H * * *` might be realized to `0 11 * * *`, or every day at the 11:00. Changing the schedule to `H H * * *` will maintain the random hour value so you'd end up with a schedule like `41 11 * * *`.

Note that there are no guarantees for schedule stability when it comes to interval schedules like `H H/6 * * *`. That is, modifying this schedule to something like `H/15 H/6 * * *` would result in a different randomized hour schedule. Though that would be a very strange schedule to use anyway.

This is because the 5 individual random field values are pregenerated regardless of whether they are used, ensuring that they're generated in the same order every time. This does not do the same for interval schedules, though there are a finite number of valid intervals for each fields so it would be possible to do the same. I'm not sure that this particular behavior is a must-have though.

### CODE REVIEW

File: cron_test.go [hunk 1]
```diff
@@ -34,53 +64,98 @@ func TestParse(t *testing.T) {
...
+
+		{"H 0 * * *", 1, testSchedule{{41}, {0}, nil, nil, nil}},
+		{"@weekly", 10, testSchedule{{14}, {16}, nil, nil, {5}}},
+		{"H H H H H", 10, testSchedule{{14}, {16}, {11}, {11}, {5}}},
+		{"H H * * H", 100, testSchedule{{43}, {8}, nil, nil, {1}}},
+		{"@monthly", 100, testSchedule{{43}, {8}, {11}, nil, nil}},
+		{"H/1 * * * *", 1, testSchedule{nil, nil, nil, nil, nil}},
+		{"* H/6 * * *", 10, testSchedule{nil, {4, 10, 16, 22}, nil, nil, nil}},
+		{"H H/6 * * *", 10, testSchedule{{14}, {4, 10, 16, 22}, nil, nil, nil}},
+		{"H/15 H/6 * * *", 10, testSchedule{{4, 19, 34, 49}, {1, 7, 13, 19}, nil, nil, nil}},
```
Comment (stefan-toubia) [inline cron_test.go:148]: These test cases describes the behavior I called out in the PR description.

File: cron.go [hunk 2]
```diff
@@ -224,10 +279,15 @@ func parseFields(expr string) (*Schedule, error) {
 	return &schedule, nil
 }
 
-func parseSinglePart(part string, fieldIndex int) (*Schedule, error) {
+var ErrParseHashedSchedule = errors.New(
+	"the \"H\" symbol cannot be used with Parse; use ParseWithHash instead",
```
Comment (cespare) [inline cron.go:283]: You can use backticks here.

File: cron.go [hunk 3]
```diff
@@ -352,3 +428,30 @@ func (s *Schedule) union(s1 *Schedule) {
...
+// newHashedFields generates random schedule values for the given seed.
+// The same schedule values are generated given the same seed.
+func newHashedFields(seed uint64) *hashedFields {
+	rng := rand.New(rand.NewSource(int64(seed)))
+	hf := &hashedFields{rng, [5]int{}}
+	for fieldIndex := 0; fieldIndex < 5; fieldIndex++ {
+		n := fieldSizes[fieldIndex]
+		if fieldIndex == 2 {
+			// Only generate random days of the month in a range of [0, 28]
+			n = 29
```
Comment (cespare) [inline cron.go:446]: This is off by 1 (the range becomes `[1, 29]` after you increment the value below).

File: cron.go [hunk 4]
```diff
@@ -352,3 +428,30 @@ func (s *Schedule) union(s1 *Schedule) {
...
+		}
+		val := rng.Intn(n)
+		switch fieldIndex {
+		// Month and day of week are 1-indexed.
+		case 2, 3:
+			val = val + 1
```
Comment (cespare) [inline cron.go:452]: You can write `val++`.

File: cron.go [hunk 5]
```diff
@@ -255,6 +318,19 @@ func parseSinglePart(part string, fieldIndex int) (*Schedule, error) {
...
+			if hf == nil {
+				return nil, ErrParseHashedSchedule
+			}
+			if !isInterval {
+				rangeStart = hf.fields[fieldIndex]
+				rangeEnd = rangeStart
+			} else {
+				// For interval schedules like H/n, generate a random start time offset
+				// between [0, n).
+				rangeStart = hf.rng.Intn(inc)
```
Comment (cespare) [inline cron.go:331]: I think there are two issues here.

First, you need to account for the 1-indexed fields here. If someone writes

    * * * H/3 *

then this might pick `rangeStart=0` which becomes -1 after `rangeStart--`, below.

Second, if someone wrote something like

    * * H/30 * *

(which is probably a mistake, but let's ignore that) then we'd want to pick `H` in [1, 28] still, I think, not [1,30]. That is, we want the minimum of whatever we'd pick otherwise and the number implied by the step interval.

File: cron_test.go [hunk 6]
```diff
@@ -34,53 +64,98 @@ func TestParse(t *testing.T) {
...
+		{"@foobar", nil},
+		{"H * * * *", ErrParseHashedSchedule},
+		{"* H/4 * * *", ErrParseHashedSchedule},
+		{"* 1,H/4 * * *", ErrParseHashedSchedule},
+		{"H(1-5) * * * *", nil},
 	} {
-		if _, err := Parse(expr); err == nil {
-			t.Fatalf("Parse accepted %q, but it is invalid", expr)
-		}
+		t.Run(tt.expr, func(t *testing.T) {
```
Comment (cespare) [inline cron_test.go:111]: Using `Run` here makes the output a lot less readable and verbose, and it also introduces some issues because the constructed "names" can include `/` which is the semantic separator for subtests. For example:

```
--- FAIL: TestParseFail (0.00s)
    --- FAIL: TestParseFail/1/3_*_*_*_* (0.00s)
        cron_test.go:114: Parse accepted "1/3 * * * *", but it is invalid
FAIL
```

I would just not use subtests here. (This also applies to the other instances of `t.Run` in this file.)

File: cron_test.go [hunk 7]
```diff
@@ -34,53 +64,98 @@ func TestParse(t *testing.T) {
...
+
+type parseHashedTestCase struct {
+	expr   string
+	seed   uint64
+	parsed testSchedule
+}
+
+func TestParseWithHash(t *testing.T) {
+	for _, testCase := range []parseHashedTestCase{
+		{"@hourly", 1, testSchedule{{41}, nil, nil, nil, nil}},
```
Comment (cespare) [inline cron_test.go:131]: Technically, the math/rand package doesn't promise to keep the same sequence of random numbers from one Go release to the next. The Go folks have talked about swapping out the PRNG for a faster one. I'm not sure they ever will because it will break lots of tests, but still -- we shouldn't depend on it here. Instead, we should use our own source of (possibly fake) randomization for testing. That has the added benefit of making it easy to test the corner cases.

File: cron.go [hunk 8]
```diff
@@ -352,3 +428,30 @@ func (s *Schedule) union(s1 *Schedule) {
...
 }
+
+type hashedFields struct {
+	rng    *rand.Rand
+	fields [5]int
+}
+
+// newHashedFields generates random schedule values for the given seed.
+// The same schedule values are generated given the same seed.
+func newHashedFields(seed uint64) *hashedFields {
```
Comment (cespare) [inline cron.go:439]: ISTM that it would simplify things to generate all the random numbers as needed rather than pre-generate five of them.

**Review by cespare — COMMENTED** [review]
Thanks Stefan. I left some comments about some issues I noticed, but in reviewing this I also found a lot of existing problems with this code that I'd like to clean up. So I think I'll merge this as-is and send a new PR that implements my suggestions as well as a bunch of other cleanup and have you review that.
